Block Entities article

natri0 opened the pull request:
Describes how to:
- [x] register & create a block entity
- [x] 'bind' it to the block w/ `BlockEntityProvider` / `BlockWithEntity`
- [x] serialize/deserialize & use custom fields in the block entity
- [x] tick

(Maybe) To do:
- [ ] custom rendering w/ `BlockEntityRenderer`
- [ ] handled screens
- [ ] networking
- [ ] data components
- [ ] container blocks

Also, idk if the article/code should be copied to 1.20.4 if there's pretty much nothing changed as far as i can tell.

natri0 committed:
Add a 'Block Entities' article

Describes how to:
- register & create a block entity
- 'bind' it to the block w/ BlockEntityProvider / BlockWithEntity
- serialize/deserialize & use custom fields in the block entity
- tick

To do:
- rendering
- handled screens
- networking
- data components

Spotless & Checkstyle checks passed.

its-miroma commented:
Thank you for your contribution!

> Also, idk if the article/code should be copied to 1.20.4 if there's pretty much nothing changed as far as i can tell.

We usually don't, but if there were no changes why not

natri0 committed:
Change the headers a bit

IMO, they were
- too long (some didn't fit in the table-of-contents)
- too similar (links back to the previous one, you couldn't see which one was "Creating and Registering the Block Entity" and which "Creating and Registering the Block")

IMB11 commented:
>  custom rendering w/ BlockEntityRenderer
 > handled screens
 > networking
 > data components
 > container blocks
 
 These can be subpages, no need to do them in this PR - this is a great start, i'll have a look at it now.

IMB11 commented:
Also, you don't need to backport changes to previous versions, the older versions are set in stone and are not really meant to be expanded upon.

IMB11 commented on develop/blocks/block-entities.md:
```diff
@@ -0,0 +1,107 @@
+# Block Entities {#block-entities}
+
+Block entities are a way to store additional data for a block, that is not part of the block state: inventory contents, custom name and so on.
+Vanilla Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
```
```suggestion
Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
```

IMB11 commented on develop/blocks/block-entities.md:
```diff
@@ -0,0 +1,107 @@
+Block entities are a way to store additional data for a block, that is not part of the block state: inventory contents, custom name and so on.
+Vanilla Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
+
+In this article, we will create a block that counts how many times it has been right-clicked.
```
```suggestion
As an example, we will create a block that counts how many times it has been right-clicked.
```

IMB11 commented on develop/blocks/block-entities.md:
```diff
@@ -0,0 +1,107 @@
+
+## Creating the Block Entity {#creating-the-block-entity}
+
+To make Minecraft "see" the new block entity, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it in the `ModBlockEntities` class, just like we did with the block.
```
```suggestion
To make Minecraft recognize and load the new block entities, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it a new `ModBlockEntities` class.
```

IMB11 commented on develop/blocks/block-entities.md:
```diff
@@ -0,0 +1,107 @@
+
+@[code transcludeWith=:::1](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/CounterBlockEntity.java)
+
+Registering a Block Entity yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:
```
```suggestion
Registering a `BlockEntity` yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:
```

IMB11 commented on develop/blocks/block-entities.md:
```diff
@@ -0,0 +1,107 @@
+
+`CounterBlockEntity::tick` is a reference to the static method `tick` we should create in the `CounterBlockEntity` class. Structuring it like this is not required, but it's a good practice to keep the code clean and organized.
+
+Let's say we want to make it so that the counter can only be incremented once every 10 ticks (2 times a second). We can do this by adding a `ticksSinceLast` field to the `CounterBlockEntity` class, and increasing it every tick:
```
Maybe mention something like "Don't forget to serialize and deserialize this new field!"

IMB11 reviewed (changes requested):
Really good structure, just a few minor changes i'd suggest.

natri0 committed:
Small changes (requested + markdownlint)

natri0 committed:
More markdownlint
